### ISSUE TITLE
Extra newline after `message()` output

### ISSUE THREAD

**Issue by gaborcsardi** [IRkernel/IRkernel#732]
- [ ] I followed the [official installation instructions](https://irkernel.github.io/installation/)
  
- [ ] I reproduced the error with the newest versions of `IRkernel`, `IRdisplay`, and `repr`
  
- [x] I included a [minimal reproducible example](https://stackoverflow.com/a/5963610/247482)
- [x] This the right repository: I think the issue is with IRkernel, not
      [IRkernel/repr](https://github.com/IRkernel/repr/issues/new) or a third party repository
  

Seems like IRkernel appends a newline character after `message()` messages:
https://github.com/IRkernel/IRkernel/blob/d0d5ccccee23d798d53b79e14c5ab5935b17f8d8/R/execution.r#L249 since https://github.com/IRkernel/IRkernel/commit/d7f0b08582c66a1b5119b50e3e6aa56a03b06d1b

This means that we cannot emit messages in pieces:
```r
{ message("foo", appendLF = FALSE); message("bar") }
```
should print 
```
foobar
```
but on Jupyter it prints
```
foo
bar

```
(With the trailing empty line at the end.)

It also means that if you use `message()` for consecutive messages, then there'll be empty lines between them.

**Comment by flying-sheep** [comment]
Hmm, I added this in d7f0b08

I think that was to make [jupyter-console](https://github.com/jupyter/jupyter_console) work properly.

Might be possible to do this differently.

**Comment by HenrikBengtsson** [comment]
I'm rediscovering what @gaborcsardi already have found, and, probably, for similar reasons; these extra newlines wreak havoc for progress frameworks relying on "same-line" message conditions for outputting progress.  It also adds lots of extra empty lines in code that outputs a lot of messages.

> Hmm, I added this in [d7f0b08](https://github.com/IRkernel/IRkernel/commit/d7f0b08582c66a1b5119b50e3e6aa56a03b06d1b)

I don't that is ideal. @flying-sheep , do you recall why you added that in the first place - what problem did it solve? There might be an alternative solution that is better.

PS. We recently ran into similar issues with RStudio, when they introduce new message and warning handlers in RStudio 2025.05 (github.com/rstudio/rstudio/issues/16038) - they have since fixed that bug. I know that Positron 2025.09 has similar problems - I haven't had time to write up a bug report.

**Comment by flying-sheep** [comment]
I assume it was because of output buffering, something like “output doesn’t get flushed if it doesn’t end in a newline”.

I assume one could change it to explicitly flush the respective output stream?

**Comment by HenrikBengtsson** [comment]
> I assume one could change it to explicitly flush the respective output stream?

Yes, I think an explicit call to a flush function is better than triggering it via extra newlines.
